flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchWriter.java

```diff
@@ -0,0 +1,326 @@
+            mailboxExecutor.yield();
+        }
+        checkpointInProgress = false;
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void close() throws Exception {
+        closed = true;
+        processor.close();
+        client.close();
+    }
+
+    private static RestClientBuilder configureRestClientBuilder(
+            RestClientBuilder builder, NetworkClientConfig networkClientConfig) {
+        if (networkClientConfig.getConnectionPathPrefix() != null) {
+            builder.setPathPrefix(networkClientConfig.getConnectionPathPrefix());
+        }
+        if (networkClientConfig.getPassword() != null
+                && networkClientConfig.getUsername() != null) {
+            final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+            credentialsProvider.setCredentials(
+                    AuthScope.ANY,
+                    new UsernamePasswordCredentials(
+                            networkClientConfig.getUsername(), networkClientConfig.getPassword()));
+            builder.setHttpClientConfigCallback(
+                    httpClientBuilder ->
+                            httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider));
+        }
+        return builder;
+    }
+
+    private static BulkProcessor configureBulkProcessor(
+            BulkProcessor.Builder builder, BulkProcessorConfig bulkProcessorConfig) {
+        // This makes flush() blocking
+        builder.setConcurrentRequests(0);
+
+        if (bulkProcessorConfig.getBulkFlushMaxActions() != -1) {
+            builder.setBulkActions(bulkProcessorConfig.getBulkFlushMaxActions());
+        }
+
+        if (bulkProcessorConfig.getBulkFlushMaxMb() != -1) {
+            builder.setBulkSize(
+                    new ByteSizeValue(bulkProcessorConfig.getBulkFlushMaxMb(), ByteSizeUnit.MB));
+        }
+
+        if (bulkProcessorConfig.getBulkFlushInterval() != -1) {
+            builder.setFlushInterval(new TimeValue(bulkProcessorConfig.getBulkFlushInterval()));
+        }
+
+        BackoffPolicy backoffPolicy;
+        if (bulkProcessorConfig.getFlushBackoffType() != null) {
+            final TimeValue backoffDelay =
+                    new TimeValue(bulkProcessorConfig.getBulkFlushBackOffDelay());
+            final int maxRetryCount = bulkProcessorConfig.getBulkFlushBackoffRetries();
+            switch (bulkProcessorConfig.getFlushBackoffType()) {
+                case CONSTANT:
+                    backoffPolicy = BackoffPolicy.constantBackoff(backoffDelay, maxRetryCount);
+                    break;
+                case EXPONENTIAL:
+                default:
+                    backoffPolicy = BackoffPolicy.exponentialBackoff(backoffDelay, maxRetryCount);
+            }
+        } else {
+            backoffPolicy = BackoffPolicy.noBackoff();
+        }
+        builder.setBackoffPolicy(backoffPolicy);
+
```

Comment:
This could be moved inside the enum's as a method `BackoffPolicy createBackoffPolicy(backoffDelay, maxRetryCount)`, especially if you add NONE.

Reply:
WDYT about making the `FlushBackoffType` similar to `DeliveryGuarantee` and configuration enum accessible to all connectors? I can imagine there a lot of connectors wanting to implement a similar thing.

 This would mean we cannot put elastic-specific code into it.

Reply:
I could even imagine that this beyond connectors and then you are right.